# Predicting sentiment from product reviews


The goal of this first notebook is to explore logistic regression and feature engineering with existing GraphLab functions.

In this notebook you will use product review data from Amazon.com to predict whether the sentiments about a product (from its reviews) are positive or negative.

* Use SFrames to do some feature engineering
* Train a logistic regression model to predict the sentiment of product reviews.
* Inspect the weights (coefficients) of a trained logistic regression model.
* Make a prediction (both class and probability) of sentiment for a new product review.
* Given the logistic regression weights, predictors and ground truth labels, write a function to compute the **accuracy** of the model.
* Inspect the coefficients of the logistic regression model and interpret their meanings.
* Compare multiple logistic regression models.

Let's get started!
    
## Fire up GraphLab Create

Make sure you have the latest version of GraphLab Create.

In [85]:
# python standard library
from __future__ import division
import math
import re
import string

# third party
import graphlab

# Data preperation

We will use a dataset consisting of baby product reviews on Amazon.com.

In [2]:
products = graphlab.SFrame('amazon_baby.gl/')

[INFO] GraphLab Create v1.8.3 started. Logging: /tmp/graphlab_server_1457292971.log


Now, let us see a preview of what the dataset looks like.

In [3]:
products

Columns:
	name	str
	review	str
	rating	float

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | 

## Build the word count vector for each review

Let us explore a specific example of a baby product.


In [4]:
products[269]

{'name': 'The First Years Massaging Action Teether',
 'rating': 5.0,
 'review': 'A favorite in our house!'}

Now, we will perform 2 simple data transformations:

1. Remove punctuation using [Python's built-in](https://docs.python.org/2/library/string.html) string functionality.
2. Transform the reviews into word-counts.

**Aside**. In this notebook, we remove all punctuations for the sake of simplicity. A smarter approach to punctuations would preserve phrases such as "I'd", "would've", "hadn't" and so forth. See [this page](https://www.cis.upenn.edu/~treebank/tokenization.html) for an example of smart handling of punctuations.

In [5]:
print(string.translate.__doc__)

translate(s,table [,deletions]) -> string

    Return a copy of the string s, where all characters occurring
    in the optional argument deletions are removed, and the
    remaining characters have been mapped through the given
    translation table, which must be a string of length 256.  The
    deletions argument is not allowed for Unicode strings.

    


In [6]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [7]:
products[269]['review'].translate(None, string.punctuation)

'A favorite in our house'

In [83]:
def remove_punctuation(text):
    return text.translate(None, string.punctuation) 

review_without_puctuation = products['review'].apply(remove_punctuation)
products['word_count'] = graphlab.text_analytics.count_words(review_without_puctuation)

Now, let us explore what the sample example above looks like after these 2 transformations. Here, each entry in the **word_count** column is a dictionary where the key is the word and the value is a count of the number of times the word occurs.

In [9]:
products[269]['word_count']

{'a': 1, 'favorite': 1, 'house': 1, 'in': 1, 'our': 1}

## Extract sentiments

We will **ignore** all reviews with *rating = 3*, since they tend to have a neutral sentiment.

In [10]:
products = products[products['rating'] != 3]
len(products)

166752

Now, we will assign reviews with a rating of 4 or higher to be *positive* reviews, while the ones with rating of 2 or lower are *negative*. For the sentiment column, we use +1 for the positive class label and -1 for the negative class label.

In [11]:
products['sentiment'] = products['rating'].apply(lambda rating : +1 if rating > 3 else -1)
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| 

Now, we can see that the dataset contains an extra column called **sentiment** which is either positive (+1) or negative (-1).

## Split data into training and test sets

Let's perform a train/test split with 80% of the data in the training set and 20% of the data in the test set. We use `seed=1` so that everyone gets the same result.

In [12]:
train_data, test_data = products.random_split(.8, seed=1)
print len(train_data)
print len(test_data)

133416
33336

# Train a sentiment classifier with logistic regression

We will now use logistic regression to create a sentiment classifier on the training data. This model will use the column **word_count** as a feature and the column **sentiment** as the target. We will use `validation_set=None` to obtain same results as everyone else.

**Note:** This line may take 1-2 minutes.

In [13]:
sentiment_model = graphlab.logistic_classifier.create(train_data,
                                                      target = 'sentiment',
                                                      features=['word_count'],
                                                      validation_set=None)

In [14]:
sentiment_model

Class                         : LogisticClassifier

Schema
------
Number of coefficients        : 121713
Number of examples            : 133416
Number of classes             : 2
Number of feature columns     : 1
Number of unpacked features   : 121712

Hyperparameters
---------------
L1 penalty                    : 0.0
L2 penalty                    : 0.01

Training Summary
----------------
Solver                        : auto
Solver iterations             : 6
Solver status                 : TERMINATED: Terminated due to numerical difficulties.
Training time (sec)           : 6.5119

Settings
--------
Log-likelihood                : inf

Highest Positive Coefficients
-----------------------------
word_count[mobileupdate]      : 41.9847
word_count[placeid]           : 41.7354
word_count[labelbox]          : 41.151
word_count[httpwwwamazoncomreviewrhgg6qp7tdnhbrefcmcrprcmtieutf8asinb00318cla0nodeid]: 40.0454
word_count[knobskeeping]      : 36.2091

Lowest Negative Coefficients
------------

**Aside**. You may get an warning to the effect of "Terminated due to numerical difficulties --- this model may not be ideal". It means that the quality metric (to be covered in Module 3) failed to improve in the last iteration of the run. The difficulty arises as the sentiment model puts too much weight on extremely rare words. A way to rectify this is to apply regularization, to be covered in Module 4. Regularization lessens the effect of extremely rare words. For the purpose of this assignment, however, please proceed with the model above.

Now that we have fitted the model, we can extract the weights (coefficients) as an SFrame as follows:

In [15]:
weights = sentiment_model.coefficients
weights.column_names()

['name', 'index', 'class', 'value', 'stderr']

There are a total of `121713` coefficients in the model. Recall from the lecture that positive weights $w_j$ correspond to weights that cause positive sentiment, while negative weights correspond to negative sentiment. 

Fill in the following block of code to calculate how many *weights* are positive ( >= 0). (**Hint**: The `'value'` column in SFrame *weights* must be positive ( >= 0)).

In [16]:
num_positive_weights = len(weights[weights['value'] >= 0])
num_negative_weights = len(weights[weights['value'] < 0])

print "Number of positive weights: %s " % num_positive_weights
print "Number of negative weights: %s " % num_negative_weights
assert num_positive_weights + num_negative_weights == 121713

Number of positive weights: 68419 
Number of negative weights: 53294 


**Quiz question:** How many weights are >= 0?
68419

## Making predictions with logistic regression

Now that a model is trained, we can make predictions on the **test data**. In this section, we will explore this in the context of 3 examples in the test dataset.  We refer to this set of 3 examples as the **sample_test_data**.

In [17]:
sample_test_data = test_data[10:13]
print sample_test_data['rating']
sample_test_data

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 3

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|   Our Baby Girl Memory Book   | Absolutely love it and all... |  5.0   |
| Wall Decor Removable Decal... | Would not purchase again o... |  2.0   |
| New Style Trailing Cherry ... | Was so excited to get this... |  1.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 2, 'all': 1, 'love... |     1     |
| {'and': 1, 'wall': 1, 'the... |     -1    |
| {'all': 1, 'money': 1, 'in... |     -1    |
+-------------------------------+-----------+
[3 rows x 5 columns]

[5.0, 2.0, 1.0]


Let's dig deeper into the first row of the **sample_test_data**. Here's the full review:

In [18]:
sample_test_data[0]['review']

'Absolutely love it and all of the Scripture in it.  I purchased the Baby Boy version for my grandson when he was born and my daughter-in-law was thrilled to receive the same book again.'

That review seems pretty positive.

Now, let's see what the next row of the **sample_test_data** looks like. As we could guess from the sentiment (-1), the review is quite negative.

In [19]:
sample_test_data[1]['review']

'Would not purchase again or recommend. The decals were thick almost plastic like and were coming off the wall as I was applying them! The would NOT stick! Literally stayed stuck for about 5 minutes then started peeling off.'

We will now make a **class** prediction for the **sample_test_data**. The `sentiment_model` should predict **+1** if the sentiment is positive and **-1** if the sentiment is negative. Recall from the lecture that the **score** (sometimes called **margin**) for the logistic regression model  is defined as:

$$
\mbox{score}_i = \mathbf{w}^T h(\mathbf{x}_i)
$$ 

where $h(\mathbf{x}_i)$ represents the features for example $i$.  We will write some code to obtain the **scores** using GraphLab Create. For each row, the **score** (or margin) is a number in the range **[-inf, inf]**.

In [20]:
scores = sentiment_model.predict(sample_test_data, output_type='margin')
print scores

[6.734619727058852, -5.734130996760089, -14.66846040446825]


### Predicting sentiment

These scores can be used to make class predictions as follows:

$$
\hat{y} = 
\left\{
\begin{array}{ll}
      +1 & \mathbf{w}^T h(\mathbf{x}_i) > 0 \\
      -1 & \mathbf{w}^T h(\mathbf{x}_i) \leq 0 \\
\end{array} 
\right.
$$

Using scores, write code to calculate $\hat{y}$, the class predictions:

In [21]:
for index, score in enumerate(scores):
    print("y_{1} = {0}".format('+1' if score > 0 else "-1", index))

y_0 = +1
y_1 = -1
y_2 = -1


Run the following code to verify that the class predictions obtained by your calculations are the same as that obtained from GraphLab Create.

In [30]:
print "Class predictions according to GraphLab Create:" 
sample_predictions = sentiment_model.predict(sample_test_data)
print(sample_predictions)

Class predictions according to GraphLab Create:
[1, -1, -1]


In [34]:
len(sample_predictions[sample_predictions == graphlab.SArray([1, -1, 1])])

2

**Checkpoint**: Make sure your class predictions match with the one obtained from GraphLab Create.

### Probability predictions

Recall from the lectures that we can also calculate the probability predictions from the scores using:
$$
P(y_i = +1 | \mathbf{x}_i,\mathbf{w}) = \frac{1}{1 + \exp(-\mathbf{w}^T h(\mathbf{x}_i))}.
$$

Using the variable **scores** calculated previously, write code to calculate the probability that a sentiment is positive using the above formula. For each row, the probabilities should be a number in the range **[0, 1]**.

In [23]:
import numpy
probability_positive = 1/(1 + numpy.exp(-scores))
print(probability_positive)

[  9.98812385e-01   3.22326818e-03   4.26155800e-07]


**Checkpoint**: Make sure your probability predictions match the ones obtained from GraphLab Create.

In [24]:
print "Class predictions according to GraphLab Create:" 
g_probability_positive =  sentiment_model.predict(sample_test_data, output_type='probability')
print(g_probability_positive)
assert all(abs(probability_positive - g_probability_positive) < 0.0000001)

Class predictions according to GraphLab Create:
[0.9988123848377193, 0.0032232681818012234, 4.261557996656561e-07]


** Quiz Question:** Of the three data points in **sample_test_data**, which one (first, second, or third) has the **lowest probability** of being classified as a positive review?

In [25]:
numpy.where(probability_positive==min(probability_positive))[0][0]

2

The third data point has the lowest probability of being classified as a positive review.

# Find the most positive (and negative) review

We now turn to examining the full test dataset, **test_data**, and use GraphLab Create to form predictions on all of the test data points for faster performance.

Using the `sentiment_model`, find the 20 reviews in the entire **test_data** with the **highest probability** of being classified as a **positive review**. We refer to these as the "most positive reviews."

To calculate these top-20 reviews, use the following steps:
1.  Make probability predictions on **test_data** using the `sentiment_model`. (**Hint:** When you call `.predict` to make predictions on the test data, use option `output_type='probability'` to output the probability rather than just the most likely class.)
2.  Sort the data according to those predictions and pick the top 20. (**Hint:** You can use the `.topk` method on an SFrame to find the top k rows sorted according to the value of a specified column.)

In [26]:
test_data['predictions'] = sentiment_model.predict(test_data, output_type='probability')

In [66]:
top_twenty = test_data.topk('predictions', 20)
for review in top_twenty:
    print("{0}\t{1}".format(review['name'], review['rating']))

Britax Decathlon Convertible Car Seat, Tiffany	4.0
Ameda Purely Yours Breast Pump - Carry All	4.0
Traveling Toddler Car Seat Travel Accessory	2.0
Shermag Glider Rocker Combo, Pecan with Oatmeal	4.0
Cloud b Sound Machine Soother, Sleep Sheep	5.0
JP Lizzy Chocolate Ice Classic Tote Set	4.0
Fisher-Price Rainforest Melodies and Lights Deluxe Gym	5.0
Lilly Gold Sit 'n' Stroll 5 in 1 Car Seat and Stroller Combination, Tuxedo Black (sunshade is not included in the offering)	5.0
Fisher-Price Deluxe Jumperoo	5.0
North States Supergate Pressure Mount Clear Choice Wood Gate	4.0
Munchkin Mozart Magic Cube	4.0
Britax Marathon Convertible Car Seat, Granite	5.0
Wizard Convertible Car Seat with LATCH in Midnight Print	5.0
Capri Stroller - Red Tech	4.0
Peg Perego Primo Viaggio Car Seat / Infant Carrier with LATCH Base - Black Sable	5.0
HALO SleepSack Micro-Fleece Wearable Blanket, Soft Pink, Small	5.0
Leachco Snoogle Total Body Pillow	5.0
Summer Infant Complete Nursery Care Kit	4.0
Safety 1st Tot-Lok F

In [95]:
top_twenty_names = [review['name'] for review in top_twenty]
def check_name(product, source=top_twenty_names):
    """
    :param:
     - `product`: regular expression to match product
     - `source`: list of names to check
    :return: True if `product` matches any of the names
    """
    for name in source:
        if re.search(product, name):
            return name
    return False

def print_check(product, source=top_twenty_names):
    found = check_name(product, source)
    if found:
        print("Found: {0}".format(found))
    return

print_check("Ameda Purely Yours")
assert check_name('Britax')
assert not check_name('umma')



Found: Ameda Purely Yours Breast Pump - Carry All


In [96]:
quiz_names = ["Snuza Portable Baby Monitor", "MamaDoo", "Britax Decathlon Convertible Car Seat, Tiffany", "Safety 1st Exchangable Tip 3 in 1 Thermometer"]
for name in quiz_names:
    print_check(name)
    

Found: Britax Decathlon Convertible Car Seat, Tiffany


**Quiz Question**: Which of the following products are represented in the 20 most positive reviews? [multiple choice]


Now, let us repeat this excercise to find the "most negative reviews." Use the prediction probabilities to find the  20 reviews in the **test_data** with the **lowest probability** of being classified as a **positive review**. Repeat the same steps above but make sure you **sort in the opposite order**.

In [93]:
bottom_twenty = test_data.topk('predictions', k=20, reverse=True)
bottom_twenty_names = [review['name'] for review in bottom_twenty]
for review in bottom_twenty:
    print("{0}\t{1}".format(review['name'], review['rating']))
    

Jolly Jumper Arctic Sneak A Peek Infant Car Seat Cover Black	5.0
Levana Safe N'See Digital Video Baby Monitor with Talk-to-Baby Intercom and Lullaby Control (LV-TW501)	1.0
Snuza Portable Baby Movement Monitor	1.0
Fisher-Price Ocean Wonders Aquarium Bouncer	2.0
VTech Communications Safe &amp; Sounds Full Color Video and Audio Monitor	1.0
Safety 1st High-Def Digital Monitor	1.0
Chicco Cortina KeyFit 30 Travel System in Adventure	1.0
Prince Lionheart Warmies Wipes Warmer	1.0
Valco Baby Tri-mode Twin Stroller EX- Hot Chocolate	1.0
Adiri BPA Free Natural Nurser Ultimate Bottle Stage 1 White, Slow Flow (0-3 months)	2.0
Munchkin Nursery Projector and Sound System, White	1.0
The First Years True Choice P400 Premium Digital Monitor, 2 Parent Unit	1.0
Nuby Natural Touch Silicone Travel Infa Feeder, Colors May Vary, 3 Ounce	1.0
Peg-Perego Tatamia High Chair, White Latte	1.0
Fisher-Price Royal Potty	1.0
Safety 1st Exchangeable Tip 3 in 1 Thermometer	1.0
Safety 1st Lift Lock and Swing Gate	1.0
Even

In [98]:
bottom_names = ["The First Years.*P400 Premium Digital Monitor", "JP Lizzy Chocolate", "Peg-Perego", "Safety 1st High-Def Digital Monitor"]
for name in bottom_names:
    print_check(name, bottom_twenty_names)
    

Found: The First Years True Choice P400 Premium Digital Monitor, 2 Parent Unit
Found: Peg-Perego Tatamia High Chair, White Latte
Found: Safety 1st High-Def Digital Monitor


**Quiz Question**: Which of the following products are represented in the 20 most negative reviews?  [multiple choice]

## Compute accuracy of the classifier

We will now evaluate the accuracy of the trained classifer. Recall that the accuracy is given by


$$
\mbox{accuracy} = \frac{\mbox{# correctly classified examples}}{\mbox{# total examples}}
$$

This can be computed as follows:

* **Step 1:** Use the trained model to compute class predictions (**Hint:** Use the `predict` method)
* **Step 2:** Count the number of data points when the predicted class labels match the ground truth labels (called `true_labels` below).
* **Step 3:** Divide the total number of correct predictions by the total number of data points in the dataset.

Complete the function below to compute the classification accuracy:

In [35]:
def get_classification_accuracy(model, data, true_labels):
    # First get the predictions
    predictions = model.predict(data)
    
    # Compute the number of correctly classified examples
    correct_classifications = len(predictions[predictions==true_labels])

    # Then compute accuracy by dividing num_correct by total number of examples
    accuracy = float(correct_classifications)/len(true_labels)
    
    return accuracy

Now, let's compute the classification accuracy of the **sentiment_model** on the **test_data**.

In [36]:
get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])

0.9145368370530358

**Quiz Question**: What is the accuracy of the **sentiment_model** on the **test_data**? Round your answer to 2 decimal places (e.g. 0.76).

0.91

**Quiz Question**: Does a higher accuracy value on the **training_data** always imply that the classifier is better?

No, it might be overfitting the training_data.

## Learn another classifier with fewer words

There were a lot of words in the model we trained above. We will now train a simpler logistic regression model using only a subet of words that occur in the reviews. For this assignment, we selected a 20 words to work with. These are:

In [37]:
significant_words = ['love', 'great', 'easy', 'old', 'little', 'perfect', 'loves', 
      'well', 'able', 'car', 'broke', 'less', 'even', 'waste', 'disappointed', 
      'work', 'product', 'money', 'would', 'return']

In [38]:
len(significant_words)

20

For each review, we will use the **word_count** column and trim out all words that are **not** in the **significant_words** list above. We will use the [SArray dictionary trim by keys functionality]( https://dato.com/products/create/docs/generated/graphlab.SArray.dict_trim_by_keys.html). Note that we are performing this on both the training and test set.

In [39]:
train_data['word_count_subset'] = train_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)
test_data['word_count_subset'] = test_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)

Let's see what the first example of the dataset looks like:

In [40]:
train_data[0]['review']

'it came early and was not disappointed. i love planet wise bags and now my wipe holder. it keps my osocozy wipes moist and does not leak. highly recommend it.'

The **word_count** column had been working with before looks like the following:

In [41]:
print train_data[0]['word_count']

{'and': 3, 'love': 1, 'it': 3, 'highly': 1, 'osocozy': 1, 'bags': 1, 'leak': 1, 'moist': 1, 'does': 1, 'recommend': 1, 'was': 1, 'wipes': 1, 'disappointed': 1, 'early': 1, 'not': 2, 'now': 1, 'holder': 1, 'wipe': 1, 'keps': 1, 'wise': 1, 'i': 1, 'planet': 1, 'my': 2, 'came': 1}


Since we are only working with a subet of these words, the column **word_count_subset** is a subset of the above dictionary. In this example, only 2 `significant words` are present in this review.

In [42]:
print train_data[0]['word_count_subset']

{'love': 1, 'disappointed': 1}


## Train a logistic regression model on a subset of data

We will now build a classifier with **word_count_subset** as the feature and **sentiment** as the target. 

In [43]:
simple_model = graphlab.logistic_classifier.create(train_data,
                                                   target = 'sentiment',
                                                   features=['word_count_subset'],
                                                   validation_set=None)
simple_model

Class                         : LogisticClassifier

Schema
------
Number of coefficients        : 21
Number of examples            : 133416
Number of classes             : 2
Number of feature columns     : 1
Number of unpacked features   : 20

Hyperparameters
---------------
L1 penalty                    : 0.0
L2 penalty                    : 0.01

Training Summary
----------------
Solver                        : auto
Solver iterations             : 6
Solver status                 : SUCCESS: Optimal solution found.
Training time (sec)           : 0.7663

Settings
--------
Log-likelihood                : 44323.7254

Highest Positive Coefficients
-----------------------------
word_count_subset[loves]      : 1.6773
word_count_subset[perfect]    : 1.5145
word_count_subset[love]       : 1.3654
(intercept)                   : 1.2995
word_count_subset[easy]       : 1.1937

Lowest Negative Coefficients
----------------------------
word_count_subset[disappointed]: -2.3551
word_count_subset[retur

SUCCESS: Optimal solution found.

+-----------+----------+--------------+-------------------+

| 6         | 7        | 0.737164     | 0.866815          |

| 5         | 6        | 0.623215     | 0.866815          |

| 4         | 5        | 0.516338     | 0.866748          |

| 3         | 4        | 0.408198     | 0.866478          |

| 2         | 3        | 0.302901     | 0.865713          |

| 1         | 2        | 0.191114     | 0.862917          |

+-----------+----------+--------------+-------------------+

| Iteration | Passes   | Elapsed Time | Training-accuracy |

+-----------+----------+--------------+-------------------+

--------------------------------------------------------

Starting Newton Method

Number of coefficients    : 21

Number of unpacked features : 20

Number of feature columns   : 1

Number of classes           : 2

Number of examples          : 133416

--------------------------------------------------------

Logistic regression:

We can compute the classification accuracy using the `get_classification_accuracy` function you implemented earlier.

In [44]:
get_classification_accuracy(simple_model, test_data, test_data['sentiment'])

0.8693004559635229

Now, we will inspect the weights (coefficients) of the **simple_model**:

In [45]:
simple_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 21

Data:
+-------------------+--------------+-------+-----------------+-----------------+
|        name       |    index     | class |      value      |      stderr     |
+-------------------+--------------+-------+-----------------+-----------------+
|    (intercept)    |     None     |   1   |   1.2995449552  | 0.0120888541331 |
| word_count_subset | disappointed |   1   |  -2.35509250061 | 0.0504149888557 |
| word_count_subset |     love     |   1   |  1.36543549368  | 0.0303546295109 |
| word_count_subset |     well     |   1   |  0.504256746398 |  0.021381300631 |
| word_count_subset |   product    |   1   | -0.320555492996 | 0.0154311321362 |
| word_count_subset |    loves     |   1   |  1.67727145556  | 0.0482328275384 |
| word_count_subset |    little    |   1   |  0.520628636025 | 0.0214691475665 |
| word_count_subset |     work     |   1   | -0.621700012425 | 0.0230330597946 |
| word_count_subset |   

Let's sort the coefficients (in descending order) by the **value** to obtain the coefficients with the most positive effect on the sentiment.

In [46]:
simple_model.coefficients.sort('value', ascending=False).print_rows(num_rows=21)

+-------------------+--------------+-------+-----------------+-----------------+
|        name       |    index     | class |      value      |      stderr     |
+-------------------+--------------+-------+-----------------+-----------------+
| word_count_subset |    loves     |   1   |  1.67727145556  | 0.0482328275384 |
| word_count_subset |   perfect    |   1   |  1.51448626703  |  0.049861952294 |
| word_count_subset |     love     |   1   |  1.36543549368  | 0.0303546295109 |
|    (intercept)    |     None     |   1   |   1.2995449552  | 0.0120888541331 |
| word_count_subset |     easy     |   1   |  1.19366189833  |  0.029288869202 |
| word_count_subset |    great     |   1   |  0.94469126948  | 0.0209509926591 |
| word_count_subset |    little    |   1   |  0.520628636025 | 0.0214691475665 |
| word_count_subset |     well     |   1   |  0.504256746398 |  0.021381300631 |
| word_count_subset |     able     |   1   |  0.191438302295 | 0.0337581955697 |
| word_count_subset |     ol

**Quiz Question**: Consider the coefficients of **simple_model**. There should be 21 of them, an intercept term + one for each word in **significant_words**. How many of the 20 coefficients (corresponding to the 20 **significant_words** and *excluding the intercept term*) are positive for the `simple_model`?

In [102]:
sum((1 for value in simple_model.coefficients['value'] if value >= 0)) - 1

10

**Quiz Question**: Are the positive words in the **simple_model** (let us call them `positive_significant_words`) also positive words in the **sentiment_model**?

In [56]:
simple_coefficients = simple_model.coefficients
sentiment_coefficients = sentiment_model.coefficients
positive_significant_words = simple_coefficients[simple_coefficients['value'] >= 0]['index']
positive_sentiment_words = sentiment_coefficients[sentiment_coefficients['value'] >=0 ]['index']
different = (word for word in positive_significant_words if word not in positive_sentiment_words)
for word in different:
    print(word)


Yes, the positive words in the simple model are also positive words in the sentiment_model.

# Comparing models

We will now compare the accuracy of the **sentiment_model** and the **simple_model** using the `get_classification_accuracy` method you implemented above.

First, compute the classification accuracy of the **sentiment_model** on the **train_data**:

In [58]:
sentiment_accuracy = get_classification_accuracy(sentiment_model, train_data, train_data['sentiment'])
print(sentiment_accuracy)

0.979440247047


Now, compute the classification accuracy of the **simple_model** on the **train_data**:

In [62]:
simple_accuracy = get_classification_accuracy(simple_model, train_data, train_data['sentiment'])
print(simple_accuracy)

0.866815074654


**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TRAINING set?

The *sentiment_model* has higher accuracy on the training set.

Now, we will repeat this excercise on the **test_data**. Start by computing the classification accuracy of the **sentiment_model** on the **test_data**:

In [60]:
sentiment_test_accuracy = get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])
print(sentiment_test_accuracy)

0.914536837053


Next, we will compute the classification accuracy of the **simple_model** on the **test_data**:

In [61]:
simple_test_accuracy = get_classification_accuracy(simple_model, test_data, test_data['sentiment'])
print(simple_test_accuracy)

0.869300455964


**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TEST set?

The *sentiment_model* also performed better on the test data.

## Baseline: Majority class prediction

It is quite common to use the **majority class classifier** as the a baseline (or reference) model for comparison with your classifier model. The majority classifier model predicts the majority class for all data points. At the very least, you should healthily beat the majority class classifier, otherwise, the model is (usually) pointless.

What is the majority class in the **train_data**?

In [63]:
num_positive  = (train_data['sentiment'] == +1).sum()
num_negative = (train_data['sentiment'] == -1).sum()
print num_positive
print num_negative

112164
21252


Now compute the accuracy of the majority class classifier on **test_data**.

**Quiz Question**: Enter the accuracy of the majority class classifier model on the **test_data**. Round your answer to two decimal places (e.g. 0.76).

In [64]:
(test_data['sentiment'] == 1).sum()/float(len(test_data))

0.8427825773938085

**Quiz Question**: Is the **sentiment_model** definitely better than the majority class classifier (the baseline)?

Yes.